File: packages/android/src/main/java/com/kozmos/components/RouteProgressRail/RouteProgressRail.kt
package com.kozmos.components.routeprogressrail

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.progressSemantics
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.kozmos.components.directionstep.DirectionType
import com.kozmos.components.directionstep.icon
import com.kozmos.tokens.KozmosThemeTokens
import com.kozmos.tokens.KozmosDimensions

/** The rail's geometry: the end dots, the travelling disc, the track. */
object KozmosRouteProgressRailGeometry {
    val dot = 10.dp
    val disc = 34.dp
    val track = KozmosDimensions.primitivesLayoutSpacing75

    fun clamp(progress: Float): Float = if (progress.isNaN()) 0f else progress.coerceIn(0f, 1f)

    /**
     * Where the disc's leading edge sits for a progress, in a rail `width`
     * wide: from just after the start dot to just before the end dot.
     */
    fun discLeading(progress: Float, width: Dp): Dp {
        val travel = (width - dot * 2 - disc).coerceAtLeast(0.dp)
        return dot + travel * clamp(progress)
    }
}

/**
 * How far along the route the visitor is, as a rail: a dot where it starts,
 * a disc carrying the current manoeuvre's arrow that travels the track, a
 * dot where it ends. TalkBack hears the label and the progress.
 */
@Composable
fun KozmosRouteProgressRail(
    progress: Float,
    type: DirectionType,
    label: String,
    modifier: Modifier = Modifier
) {
    val clamped = KozmosRouteProgressRailGeometry.clamp(progress)
    BoxWithConstraints(
        modifier = modifier
            .fillMaxWidth()
            .height(KozmosRouteProgressRailGeometry.disc)
            .progressSemantics(clamped)
            .semantics { contentDescription = label }
    ) {
        Box(
            modifier = Modifier
                .align(Alignment.CenterStart)
                .padding(horizontal = KozmosRouteProgressRailGeometry.dot)
                .fillMaxWidth()
                .height(KozmosRouteProgressRailGeometry.track)
                .background(KozmosThemeTokens.primitivesColorsBackground300, CircleShape)
        )
        Box(
            modifier = Modifier
                .align(Alignment.CenterStart)
                .size(KozmosRouteProgressRailGeometry.dot)
                .background(KozmosThemeTokens.primitivesColorsTheme500, CircleShape)
        )
        Box(
            modifier = Modifier
                .align(Alignment.CenterEnd)
                .size(KozmosRouteProgressRailGeometry.dot)
                .background(KozmosThemeTokens.primitivesColorsBackground300, CircleShape)
        )
        Box(
            modifier = Modifier
                .align(Alignment.CenterStart)
                .offset(x = KozmosRouteProgressRailGeometry.discLeading(clamped, maxWidth))
                .size(KozmosRouteProgressRailGeometry.disc)
                .background(KozmosThemeTokens.primitivesColorsTheme500, CircleShape),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                imageVector = type.icon(),
                contentDescription = null,
                tint = KozmosThemeTokens.primitivesColorsBackground0,
                modifier = Modifier.size(KozmosDimensions.primitivesLayoutSizing300)
            )
        }
    }
}

File: packages/ios/Sources/Components/RouteProgressRail/RouteProgressRail.swift
import SwiftUI

/// How far along the route the visitor is, as a rail: a dot where it starts,
/// a disc carrying the current manoeuvre's arrow that travels the track, a dot
/// where it ends. Mirrors the product prototype's rail — 10-point dots, a
/// 34-point theme disc, a 6-point track.
///
/// VoiceOver hears the label the caller gives ("Step 2 of 4") and the
/// progress as a percentage; the rail updates as the route is walked.
public struct KozmosRouteProgressRail: View {
    let progress: Double
    let type: DirectionType
    let label: String

    public init(progress: Double, type: DirectionType, label: String) {
        self.progress = progress
        self.type = type
        self.label = label
    }

    static let dot: CGFloat = 10
    static let disc: CGFloat = 34
    static let track: CGFloat = 6

    /// Where the disc's leading edge sits for a progress, in a rail `width`
    /// wide: from just after the start dot to just before the end dot.
    static func discLeading(progress: Double, width: CGFloat) -> CGFloat {
        let clamped = CGFloat(min(max(progress, 0), 1))
        let travel = max(width - dot * 2 - disc, 0)
        return dot + travel * clamped
    }

    private var clamped: Double { min(max(progress, 0), 1) }

    public var body: some View {
        GeometryReader { geometry in
            let width = geometry.size.width
            ZStack(alignment: .leading) {
                Capsule()
                    .fill(KozmosColors.primitivesColorsBackground300)
                    .frame(width: max(width - Self.dot * 2, 0), height: Self.track)
                    .offset(x: Self.dot)
                Circle()
                    .fill(KozmosColors.primitivesColorsTheme500)
                    .frame(width: Self.dot, height: Self.dot)
                Circle()
                    .fill(KozmosColors.primitivesColorsBackground300)
                    .frame(width: Self.dot, height: Self.dot)
                    .offset(x: max(width - Self.dot, 0))
                Circle()
                    .fill(KozmosColors.primitivesColorsTheme500)
                    .frame(width: Self.disc, height: Self.disc)
                    .overlay(
                        Image(systemName: type.iconName)
                            .font(.system(size: 15, weight: .semibold))
                            .foregroundColor(KozmosColors.primitivesColorsBackground0)
                    )
                    .offset(x: Self.discLeading(progress: clamped, width: width))
            }
            // Leading, not centred: the stack is only as wide as its widest
            // child, the track, and centring it would shift the rail by a dot.
            .frame(width: width, height: Self.disc, alignment: .leading)
        }
        .frame(height: Self.disc)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(label)
        .accessibilityValue("\(Int((clamped * 100).rounded())) percent")
        .accessibilityAddTraits(.updatesFrequently)
    }
}
